microprofile/WFLY-15832-microconfig-config-root-folder.adoc

```diff
@@ -0,0 +1,99 @@
+= WFLY-15832 - MP Config: Ability to specify root config source directory
+:author:            Kabir Khan
+:email:             kkhan@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+The MicroProfile Config with SmallRye subsystem currently supports configuration of several types of Config Sources. One of these allows you to specify a directory, wherein each file is used as a config property. The file name is the name of the property, and the contents of the file is the value of the property.
+
+This is useful on OpenShift where ConfigMaps, the ServiceBinding opertator (and probably other things) expose config via this mechanism for use in applications.
+
+This RFE is about enhancing the subsystem to allow specifying a root directory, of which the top-level directories under it will appear as Comfig Sources. The motivation is that each 'thing' on OpenShift installs their config under a root directory. e.g. in the ServiceBinding operator case this is under the `/bindings` directory and ConfigMaps are installed under `/etc/config`.
+
+So if you have two ConfigMaps called `cfg-a` and `cfg-b`, their respective directories containing the config files will be `/etc/config/cfg-a` and `/etc/config/cfg-b`. At present this means we need to add two Config Source entries to the subsystem, and we need to know the exact names of both, in order to pick these up.
+
+While we will keep the existing mechanism, the proposal is to be able to specify a directory as a root for configuration, and that each top level directory becomes a Config Source as with the previous mechanism.
+
+This has the advantage of only needing to specify one Config Source root in order to pick up all ConfigMaps, one to pick up all ServiceBinding configs, and so on. An added bonus is not needing to know the exact names of the configuration directories, which are in some cases autogenerated.
+
+From the previous example, specifying `/etc/config` as a *root* ConfigSource will pick up the configs from both the `/etc/config/cfg-a` and `/etc/config/cfg-b` directories.
+
+== Issue Metadata
+
+=== Issue
+
+https://issues.redhat.com/browse/WFLY-15832[WFLY-15832 - MP Config: Ability to specify root config source directory]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1783[EAP7-1783 - MP Config: Ability to specify root config source directory]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+All code changes are in the MicroProfile Config with SmallRye subsystem in WildFly.
```

Comment:
For me to understand. This feature goes beyond what MicroProfile Specification mandates to be MP compliant?

Reply:
@mchoma Yes, as does the existing feature it is improving.

The spec only requires us to provide ConfigSources backed by system properties and env vars and any META-INF/ microprofile-config.properties found on the classpath. 

https://download.eclipse.org/microprofile/microprofile-config-3.0.1/microprofile-config-spec-3.0.1.html#default_configsources

Note that the config_ordinal file discussed below is a mechanism to provide the spec behavior from https://download.eclipse.org/microprofile/microprofile-config-3.0.1/microprofile-config-spec-3.0.1.html#_manually_defining_the_ordinal_of_a_built_in_configsource